libbpf-tools/ttysnoop.c

```diff
@@ -0,0 +1,313 @@
+			argp_usage(state);
+		}
+
+		char path[4096] = {};
+		struct stat st;
+
+		if (arg[0] != '/') {
+			strcpy(path, "/dev/pts/");
+			strcat(path, arg);
+		} else {
+			strcpy(path, arg);
+		}
+
+		if (stat(path, &st)) {
+			warning("Failed to stat console file: %s\n", arg);
+			argp_usage(state);
+		}
+		env.pts_inode = st.st_ino;
+		break;
+	}
+	case ARGP_KEY_END:
+		if (env.pts_inode == -1)
+			argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static int libbpf_print_fn(enum libbpf_print_level level, const char *format,
+			   va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static void sig_handler(int sig)
+{
+	exiting = 1;
+}
+
+static int handle_event(void *ctx, void *data, size_t data_sz)
+{
+	const struct event *e = data;
+	char buf[BUFSIZ+1] = {};
+
+	memcpy(buf, e->buf, e->count);
+	printf("%s", buf);
+	fflush(stdout);
+
+	return 0;
+}
+
+static void handle_lost_events(void *ctx, int cpu, __u64 lost_cnt)
+{
+	warning("Lost %llu event on CPU #%d!\n", lost_cnt, cpu);
+}
+
+static bool fallback_to_compare_kernel_version(void)
+{
+	struct utsname sys_info;
+	int major1, minor1, patch1;
+	int major2, minor2, patch2;
+	const char *version = "5.10.11";
+
+	uname(&sys_info);
+
+	sscanf(sys_info.release, "%d.%d.%d%*s", &major1, &minor1, &patch1);
+	sscanf(version, "%d.%d.%d%*s", &major2, &minor2, &patch2);
+
+	if (major1 < major2)
+		return false;
+	else if (major1 > major2)
+		return true;
+
+	if (minor1 < minor2)
+		return false;
+	else if (minor1 > minor2)
+		return true;
+
+	if (patch1 < patch2)
+		return false;
+	else if (patch1 > patch2)
+		return true;
+
+	return false;
+}
+
+static bool tty_write_is_newly(void)
+{
+	const struct btf_param *param;
+	const struct btf_type *type;
+	__s32 id;
+	const char *name;
+	struct btf *btf;
+
+	btf = btf__load_vmlinux_btf();
+	if (!btf) {
+		warning("No BTF, cannot determine type info: %s", strerror(errno));
+		goto failed;
+	}
+
+	id = btf__find_by_name_kind(btf, "tty_write", BTF_KIND_FUNC);
+	if (id <= 0) {
+		warning("Can't find function tty_write in BTF: %s\n",
+			strerror(-id));
+		goto failed;
+	}
+
+	type = btf__type_by_id(btf, id);
+	if (!type || BTF_INFO_KIND(type->info) != BTF_KIND_FUNC)
+		goto failed;
+
+	type = btf__type_by_id(btf, type->type);
+	if (!type || BTF_INFO_KIND(type->info) != BTF_KIND_FUNC_PROTO)
+		goto failed;
+
+	/* try to find "struct kiocb *" or "struct file *" */
+	param = (struct btf_param *)(type + 1);
+	type = btf__type_by_id(btf, param->type);
+	if (!type || BTF_INFO_KIND(type->info) != BTF_KIND_PTR)
+		goto failed;
+
+	type = btf__type_by_id(btf, type->type);
+	if (!type || BTF_INFO_KIND(type->info) != BTF_KIND_STRUCT)
+		goto failed;
+
+	name = btf__str_by_offset(btf, type->name_off);
+	if (strcmp(name, "kiocb") != 0)
+		return false;
+
+	return true;
+
+failed:
+	return fallback_to_compare_kernel_version();
```

Comment:
This fallback seems redundant. If there are something wrong with BTF, the tool itself can't run either.

Reply:
No, for example, on Arm64, the tty_write function cannot be read using BTF. At this time, it can only be run through function version matching, otherwise it cannot run on arm64